README.md
# promise-any-ponyfill

[![Build Status][ci-img]][ci]
[![BrowserStack Status][browserstack-img]][browserstack]

[`Promise.any`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any)
[ponyfill](https://ponyfill.com).

> `Promise.any()` takes an iterable of Promise objects and, as soon as one of
> the promises in the iterable fulfills, returns a single promise that resolves
> with the value from that promise. If no promises in the iterable fulfill (if
> all of the given promises are rejected), then the returned promise is rejected
> with an `AggregateError`, a new subclass of `Error` that groups together
> individual errors. Essentially, this method is the opposite of
> `Promise.all()`.

## Install

```sh
npm install promise-any-ponyfill --save
```

## Usage

```js
import pAny from 'promise-any-ponyfill';

(async () => {
	try {
		const first = await pAny([
			Promise.resolve('becky'),
			Promise.resolve('roxy'),
			Promise.resolve('sadie')
		]);
		// Any of the promises was fulfilled.
		console.log(first);
		// → 'becky'
	} catch (error) {
		// All of the promises were rejected.
		console.log(error);
	}
})();
```

You can **use named export `preferNative` if you wish to use native
implementation if it’s available**. In all other cases, ponyfill will be used.
Beware of
[caveats](https://github.com/sindresorhus/ponyfill#:~:text=Ponyfills%20should,underlying%20environment)!

## API

### any(iterable)

Returns:

-   An **already rejected** `Promise` if the iterable passed is empty.
-   An **asynchronously resolved** `Promise` if the iterable passed contains no
    promises.
-   A **pending** `Promise` in all other cases. This returned promise is then
    resolved/rejected **asynchronously** (as soon as the stack is empty) when
    any of the promises in the given iterable resolve, or if all the promises
    have rejected.

#### iterable

An iterable object, such as an `Array`.

## Browser support

Tested in Chrome 90, Firefox 88, Internet Explorer 11 and should work in all
modern browsers
([support based on Browserslist configuration](https://browserslist.dev/?q=bGFzdCAyIHZlcnNpb25zLCBub3QgaWUgMTAsIGllID49IDEx)).

Assumes `Promise` and `AggregateError` are polyfilled or available in global
context.

## Acknowledgments

-   [Proposed implementation and description on StackOverflow](https://stackoverflow.com/a/37235274/178058)
-   [Proposed implementatio on TC39 feature repository](https://github.com/tc39/proposal-promise-any/issues/6#issue-404950354)

## Related

-   [`AggregateError` ponyfill](https://github.com/niksy/aggregate-error-ponyfill)

## Test

Test suite is taken and modified from
[es-shims](https://github.com/es-shims/Promise.any/blob/master/test/tests.js)
test suite.

For automated tests, run `npm run test:automated` (append `:watch` for watcher
support).

## License

MIT © [Ivan Nikolić](http://ivannikolic.com)

<!-- prettier-ignore-start -->

[ci]: https://travis-ci.com/niksy/promise-any-ponyfill
[ci-img]: https://travis-ci.com/niksy/promise-any-ponyfill.svg?branch=master
[browserstack]: https://www.browserstack.com/
[browserstack-img]: https://www.browserstack.com/automate/badge.svg?badge_key=MEtwckxKSGJ6eDE1UTg5YTRNUEM2Um50K0ZDVjFhUGpnNlE3THFUVHE5TT0tLTdzYkVjV05KZ0tVYUJuTVNHMG55blE9PQ==--3c1c4b767baf981e712844c150fb667befa43547

<!-- prettier-ignore-end -->


CHANGELOG.md
# Changelog

## [Unreleased][]

## [1.0.0][] - 2021-06-16

### Added

-   Initial implementation

[unreleased]:
	https://github.com/niksy/promise-any-ponyfill/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/niksy/promise-any-ponyfill/tree/v1.0.0


test/index.js
/* globals globalThis:false */

import assert from 'assert';
import { preferNative as AggregateError } from 'aggregate-error-ponyfill';
import any, { preferNative } from '../index';

if (!globalThis.AggregateError) {
	globalThis.AggregateError = AggregateError;
}

before(function () {
	window.fixture.load('/test/fixtures/index.html');
});

after(function () {
	window.fixture.cleanup();
});

const Subclass = (function () {
	try {
		// eslint-disable-next-line no-new-func
		return new Function(
			'class Subclass extends Promise { constructor(...args) { super(...args); this.thenArgs = []; } then(...args) { Subclass.thenArgs.push(args); this.thenArgs.push(args); return super.then(...args); } } Subclass.thenArgs = []; return Subclass;'
		)();
	} catch (error) {
		/**/
	}

	return false;
})();

const assertArray = (value, length, assertType) => {
	assert.ok(Array.isArray(value), 'value is an array');
	assert.equal(value.length, length, `length is ${length}`);
	if (typeof assertType === 'function') {
		for (let index = 0; index < value.length; index += 1) {
			assertType(value[index]);
		}
	}
};

const a = {};
const b = {};
const c = {};

it('should handle empty iterable', async function () {
	try {
		await any([]);
	} catch (error) {
		assert.equal(
			error instanceof AggregateError,
			true,
			'is an AggregateError'
		);
		assert.deepEqual(error.errors, []);
	}
});

it('should handle no promise values', async function () {
	const result = await any([a, b, c]);
	assert.deepEqual(result, a);
});

it('should handle all fulfilled', async function () {
	const result = await any([
		Promise.resolve(a),
		Promise.resolve(b),
		Promise.resolve(c)
	]);
	assert.deepEqual(result, a);
});

it('should handle all rejected', async function () {
	try {
		await any([Promise.reject(a), Promise.reject(b), Promise.reject(c)]);
	} catch (error) {
		assert.equal(
			error instanceof AggregateError,
			true,
			'is an AggregateError'
		);
		assert.deepEqual(error.errors, [a, b, c]);
	}
});

it('should handle mixed', async function () {
	const mixed = await any([a, Promise.resolve(b), Promise.reject(c)]);
	assert.deepEqual(mixed, a);

	const onlyPromises = await any([
		Promise.reject(a),
		Promise.resolve(b),
		Promise.reject(c)
	]);
	assert.deepEqual(onlyPromises, b);
});

it('should handle poisoned .then', async function () {
	const poison = new EvalError();
	const promise = new Promise(function () {});
	promise.then = function () {
		throw poison;
	};
	try {
		await any([promise]);
	} catch (error) {
		assert.equal(
			error instanceof AggregateError,
			true,
			'error is an AggregateError'
		);
		assert.deepEqual(
			error.errors,
			[poison],
			'rejection showed up as expected'
		);
	}
});

it('should use native implementation if it’s available', async function () {
	const result = await preferNative([a, b, c]);
	assert.deepEqual(result, a);
});

const isIE = window.navigator.msPointerEnabled;
const conditionalDescribe = isIE ? describe.skip : describe;

conditionalDescribe('Inheritance', function () {
	it('should preserve correct subclass', function () {
		const promise = any.call(Subclass, [1]);
		assert.ok(
			promise instanceof Subclass,
			'promise is instanceof Subclass'
		);
		assert.equal(
			promise.constructor,
			Subclass,
			'promise.constructor is Subclass'
		);
	});

	it('should invoke the subclass’ then', function () {
		Subclass.thenArgs.length = 0;

		const original = Subclass.resolve();
		assertArray(Subclass.thenArgs, 0);
		assertArray(original.thenArgs, 0);

		any.call(Subclass, [original]);

		assertArray(original.thenArgs, 1);
		assertArray(Subclass.thenArgs, 3);
	});
});


karma.conf.js
'use strict';

const path = require('path');
const fs = require('fs');
const { default: resolve } = require('@rollup/plugin-node-resolve');
const commonjs = require('@rollup/plugin-commonjs');
const nodeBuiltins = require('rollup-plugin-node-builtins');
const globals = require('rollup-plugin-node-globals');
const { default: babel } = require('@rollup/plugin-babel');
const istanbul = require('rollup-plugin-istanbul');
const rollupConfig = require('./rollup.config');

let config;

const isCI =
	typeof process.env.CI !== 'undefined' && process.env.CI !== 'false';
const isPR =
	typeof process.env.TRAVIS_PULL_REQUEST !== 'undefined' &&
	process.env.TRAVIS_PULL_REQUEST !== 'false';
const local = !isCI || (isCI && isPR);

const port = 0;

if (local) {
	config = {
		browsers: ['Chrome']
	};
} else {
	config = {
		hostname: 'bs-local.com',
		browserStack: {
			username: process.env.BROWSER_STACK_USERNAME,
			accessKey: process.env.BROWSER_STACK_ACCESS_KEY,
			startTunnel: true,
			project: 'promise-any-ponyfill',
			name: 'Automated (Karma)',
			build: 'Automated (Karma)'
		},
		customLaunchers: {
			'BS-Chrome': {
				'base': 'BrowserStack',
				'project': 'promise-any-ponyfill',
				'build': 'Automated (Karma)',
				'browser': 'Chrome',
				'browser_version': '90',
				'name': 'Chrome',
				'os': 'Windows',
				'os_version': '7'
			},
			'BS-Firefox': {
				'base': 'BrowserStack',
				'project': 'promise-any-ponyfill',
				'build': 'Automated (Karma)',
				'browser': 'Firefox',
				'browser_version': '88',
				'name': 'Firefox',
				'os': 'Windows',
				'os_version': '7'
			},
			'BS-IE': {
				'base': 'BrowserStack',
				'project': 'promise-any-ponyfill',
				'build': 'Automated (Karma)',
				'browser': 'IE',
				'browser_version': '11',
				'name': 'IE',
				'os': 'Windows',
				'os_version': '7'
			}
		},
		browsers: ['BS-Chrome', 'BS-Firefox', 'BS-IE']
	};
}

module.exports = function (baseConfig) {
	baseConfig.set({
		basePath: '',
		frameworks: ['mocha', 'fixture'],
		files: ['test/**/*.html', { pattern: 'test/**/*.js', watched: false }],
		exclude: [],
		preprocessors: {
			'test/**/*.html': ['html2js'],
			'test/**/*.js': ['rollup', 'sourcemap']
		},
		reporters: ['mocha', 'coverage-istanbul'],
		port: port,
		colors: true,
		logLevel: baseConfig.LOG_INFO,
		autoWatch: false,
		client: {
			captureConsole: true
		},
		browserConsoleLogOptions: {
			level: 'log',
			format: '%b %T: %m',
			terminal: true
		},
		rollupPreprocessor: {
			plugins: [
				nodeBuiltins(),
				babel({
					exclude: 'node_modules/**',
					babelHelpers: 'runtime'
				}),
				resolve({
					preferBuiltins: true
				}),
				commonjs(),
				babel({
					include: 'node_modules/{has-flag,supports-color}/**',
					babelHelpers: 'runtime',
					babelrc: false,
					configFile: path.resolve(__dirname, '.babelrc')
				}),
				globals(),
				...rollupConfig.plugins.filter(
					({ name }) => !['babel'].includes(name)
				),
				istanbul({
					exclude: ['test/**/*.js', 'node_modules/**/*']
				})
			],
			output: {
				format: 'iife',
				name: 'promiseAnyPonyfill',
				sourcemap: baseConfig.autoWatch ? false : 'inline', // Source map support has weird behavior in watch mode
				intro: 'window.TYPED_ARRAY_SUPPORT = false;' // IE9
			}
		},
		coverageIstanbulReporter: {
			dir: path.join(__dirname, 'coverage/%browser%'),
			fixWebpackSourcePaths: true,
			reports: ['html', 'text'],
			thresholds: {
				global: JSON.parse(
					fs.readFileSync(path.join(__dirname, '.nycrc'), 'utf8')
				)
			}
		},
		singleRun: true,
		concurrency: Infinity,
		...config
	});
};
